core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h

```diff
@@ -37,79 +37,115 @@ void secp256k1_default_error_callback_fn(const char *str, void *data) {
-static void *secp256k1_ctx_buf = NULL;
-static size_t secp256k1_ctx_size = 0;
+/// package: trezorcrypto.secp256k1_zkp
 
-static const secp256k1_context *mod_trezorcrypto_secp256k1_context_create(
-    void) {
-  if (secp256k1_ctx == NULL) {
-    secp256k1_ctx_size = secp256k1_context_preallocated_size(
-        SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
-    secp256k1_ctx_buf = m_new_maybe(uint8_t, secp256k1_ctx_size);
-    if (!secp256k1_ctx_buf) {
-      mp_raise_ValueError("secp256k1_zkp context is too large");
-    }
-    secp256k1_ctx = secp256k1_context_preallocated_create(
-        secp256k1_ctx_buf, SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+/// class Context:
+///     """
+///     Owns a secp256k1 context.
+///     Can be allocated once and re-used between subsequent operations.
+///     """
+///
+typedef struct _mp_obj_secp256k1_context_t {
+  mp_obj_base_t base;
+  secp256k1_context *secp256k1_ctx;
+  size_t secp256k1_ctx_size;
+  uint8_t secp256k1_ctx_buf[0];  // to be allocate via m_new_obj_var_maybe().
+} mp_obj_secp256k1_context_t;
 
-    uint8_t rand[32];
-    random_buffer(rand, 32);
-    int ret = secp256k1_context_randomize(secp256k1_ctx, rand);
-    if (ret != 1) {
-      mp_raise_msg(&mp_type_RuntimeError, "secp256k1_context_randomize failed");
-    }
+/// def __init__(self):
+///     """
+///     Allocate and initialize secp256k1_context.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_make_new(
+    const mp_obj_type_t *type, size_t n_args, size_t n_kw,
+    const mp_obj_t *args) {
+  mp_arg_check_num(n_args, n_kw, 0, 0, false);
+
+  const size_t secp256k1_ctx_size = secp256k1_context_preallocated_size(
+      SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+  mp_obj_secp256k1_context_t *o = m_new_obj_var_maybe(
+      mp_obj_secp256k1_context_t, uint8_t, secp256k1_ctx_size);
+  if (!o) {
+    mp_raise_ValueError("secp256k1_zkp context is too large");
+  }
+  o->base.type = type;
+  o->secp256k1_ctx_size = secp256k1_ctx_size;
+  o->secp256k1_ctx = secp256k1_context_preallocated_create(
+      o->secp256k1_ctx_buf, SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+  uint8_t rand[32] = {0};
+  random_buffer(rand, 32);
+  int ret = secp256k1_context_randomize(o->secp256k1_ctx, rand);
+  if (ret != 1) {
+    mp_raise_msg(&mp_type_RuntimeError, "secp256k1_context_randomize failed");
   }
-  return secp256k1_ctx;
+
+  return MP_OBJ_FROM_PTR(o);
 }
 
-STATIC void mod_trezorcrypto_secp256k1_context_delete(void) {
-  secp256k1_context_preallocated_destroy(secp256k1_ctx);
-  m_del(uint8_t, secp256k1_ctx_buf, secp256k1_ctx_size);
-  secp256k1_ctx_buf = NULL;
-  secp256k1_ctx = NULL;
+/// def __del__(self):
+///     """
+///     Destructor.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context___del__(mp_obj_t self) {
+  mp_obj_secp256k1_context_t *o = MP_OBJ_TO_PTR(self);
+  secp256k1_context_preallocated_destroy(o->secp256k1_ctx);
```

Comment:
I guess we need to call `m_del(uint8_t, o->secp256k1_ctx_buf, o->secp256k1_ctx_size);` after the destroy?

Reply:
I am not sure... 
IIUC, `m_new_obj_var_maybe` is calling `m_alloc` under the hood to allocate extra memory for the [flexible array member `secp256k1_ctx_buf`](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html) at the end of `mp_obj_secp256k1_context_t`:
https://github.com/trezor/micropython/blob/c5d8107e2af090ff4547851297a484652fde4242/py/misc.h#L63
So I think that the memory allocated for `secp256k1_ctx_buf` should be deallocated together with the surrounding `mp_obj_secp256k1_context_t` struct, but I may be wrong.

Reply:
Ah, correct. I misread the code.